sessionctx/variable/sysvar.go

```diff
@@ -810,23 +810,35 @@ var defaultSysVars = []*SysVar{
 	{Scope: ScopeGlobal, Name: InitConnect, Value: ""},
 
 	/* TiDB specific variables */
+	{Scope: ScopeGlobal | ScopeSession, Name: TiDBEnableLocalTxn, Value: BoolToOnOff(DefTiDBEnableLocalTxn), Hidden: true, Type: TypeBool, SetSession: func(s *SessionVars, val string) error {
+		s.EnableLocalTxn = TiDBOptOn(val)
+		return nil
+	}},
 	// TODO: TiDBTxnScope is hidden because local txn feature is not done.
 	{Scope: ScopeSession, Name: TiDBTxnScope, skipInit: true, Hidden: true, Value: func() string {
-		if isGlobal, _ := config.GetTxnScopeFromConfig(); isGlobal {
+		if txnScope := config.GetTxnScopeFromConfig(); txnScope == kv.GlobalTxnScope {
 			return kv.GlobalTxnScope
 		}
+		// If the zone label is configured, we set @@txn_scope as LocalTxnScope by default.
 		return kv.LocalTxnScope
 	}(), SetSession: func(s *SessionVars, val string) error {
 		switch val {
 		case kv.GlobalTxnScope:
 			s.TxnScope = kv.NewGlobalTxnScopeVar()
 		case kv.LocalTxnScope:
-			s.TxnScope = kv.GetTxnScopeVar()
+			if !s.EnableLocalTxn {
+				return ErrWrongValueForVar.GenWithStack("@@txn_scope can not be set when @@tidb_enable_local_txn is off")
+			}
+			s.TxnScope = kv.NewLocalTxnScopeVar(config.GetTxnScopeFromConfig())
 		default:
 			return ErrWrongValueForVar.GenWithStack("@@txn_scope value should be global or local")
 		}
 		return nil
 	}, GetSession: func(s *SessionVars) (string, error) {
+		// Always return the Global TxnScope if not enable Local Txn.
+		if !s.EnableLocalTxn {
+			return kv.GlobalTxnScope, nil
+		}
```

Comment:
I suggest not adding this logic as it may mislead us by covering the real value of `txn_scope`.

Reply:
I changed a way to do that. PTAL.